### ISSUE TITLE
csidriver: support multiple edge nodes /multiple csi drivers per edge node

### ISSUE THREAD

**Issue by moolen** [kubeedge/kubeedge#3471]
related: original [csi design doc](https://github.com/container-storage-interface/spec/blob/master/lib/go/csi/csi.pb.go#L802)

tl;dr
* csidriver needs to be able to send `Volume*` requests to specific edge nodes
* we can use the CSI topology feature to address specific nodes
* csidriver needs a way to support multiple csi drivers (like hostpath, nfs etc.) on a given edge node.

**Why is this needed**:
`csidriver` in its current form supports only a single node (specified via `--node-id`). csidriver should be smarter and provision volumes on many edge nodes. There is one central problem to solve: **how do we know where a request should go to**?

We can do this by using `volumeBindingMode=WaitForFirstConsumer` while attaching correct labels to `Kind=Node` and configuring `Kind=CSINode` correctly.

Example storage class: We must enable `volumeBindingMode=WaitForFirstConsumer` to delay volume binding until a pod was  scheduled on a edge node. Now we know where the pod lives and create a volume there.

```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: csi-kubeedge
provisioner: kubeedge-csidriver
volumeBindingMode: WaitForFirstConsumer
parameters:
  edgeDriverName: csi-hostpath # <- used to select the correct csi driver on the edge, see below
```

external-provisioner uses `Kind=CSINode` to populate the `csi.CreateVolumeRequest` ([spec](https://github.com/container-storage-interface/spec/blob/master/lib/go/csi/csi.pb.go#L863-L874)) with `AccessibilityRequirements` [source](https://github.com/kubernetes-csi/external-provisioner/blob/master/pkg/controller/topology.go#L131-L134). That means for us we need to set up the CSINodes correctly in order to integrate correctly with external-provisioner.

```yaml
apiVersion: storage.k8s.io/v1
kind: CSINode
metadata:
  name: edge-1
spec:
  drivers:
  - name: csi-hostpath
    nodeID: edge-1
    topologyKeys:
    - csi.kubeedge.io/nodeid # <- this is something we can set from edged
    - topology.hostpath.csi/node
```
Further, external-provisioner relies on correct `Kind=Node` labels (the one specified above in the `topologyKeys`).

```yaml
apiVersion: v1
kind: Node
metadata:
  labels:
    csi.kubeedge.io/nodeid: edge-1 # <-- this is required and must be set by the operator.
    kubernetes.io/arch: amd64
    kubernetes.io/hostname: edge-1
    kubernetes.io/os: linux
    node-role.kubernetes.io/agent: ""
    node-role.kubernetes.io/edge: ""
```

How do we support multiple csi drivers **on the edge**?

* define multiple storage classes: e.g. one for hostpath, one for nfs.
* a user has to specify `parameter.edgeDriverName` to select the correct driver on the edge
* TODO: we need to find a way to pass that metadata down to the edge, we can not simply serialize the `Volume*` requests anymore.

## Alternatives

* run all csi-related controllers on edge nodes. That we we circumvent the routing problem in `csidriver` on the cloud side
* implement our own external-provisioner that is aware of the edge nodes so we don't need to rely on topology

CC @lvchenggang @fisherxu @gy95

**Comment by shumintao** [comment]
When I use your CSI branch (Topo-IMPLL), the pod is still pending and I find cloudcore reporting this error. Here is my Cloudcore role profile.

```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: cloudcore
  labels:
    k8s-app: kubeedge
    kubeedge: cloudcore
rules:
- apiGroups: [""]
  resources: ["nodes", "nodes/status", "configmaps", "pods", "pods/status", "secrets", "endpoints", "services","persistentvolumes","persistentvolumeclaims"]
  verbs: ["get", "list", "watch", "create", "update"]
- apiGroups: [""]
  resources: ["namespaces"]
  verbs: ["get", "create"]
- apiGroups: [""]
  resources: ["pods/status"]
  verbs: ["patch"]
- apiGroups: [""]
  resources: ["pods"]
  verbs: ["delete"]
- apiGroups: ["coordination.k8s.io"]
  resources: ["leases"]
  verbs: ["get", "list", "watch", "create", "update"]
- apiGroups: ["devices.kubeedge.io"]
  resources: ["devices", "devicemodels", "devices/status", "devicemodels/status"]
  verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
- apiGroups: ["reliablesyncs.kubeedge.io"]
  resources: ["objectsyncs", "clusterobjectsyncs", "objectsyncs/status", "clusterobjectsyncs/status"]
  verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
- apiGroups: ["rules.kubeedge.io"]
  resources: ["rules", "ruleendpoints", "rules/status", "ruleendpoints/status"]
  verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
- apiGroups: ["networking.istio.io"]
  resources: ["virtualservices", "destinationrules", "serviceentries", "gateways"]
  verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
- apiGroups: ["storage.k8s.io"]
  resources: ["csinodes"]
  verbs: ["get", "list", "watch", "create", "update", "patch"]
- apiGroups: ["storage.k8s.io"]
  resources: ["volumeattachments"]
  verbs: ["get", "list", "watch"]
```

- cloudcore yaml:

```
E0114 08:12:12.744876       1 objectsync.go:36] failed to get obj(gvr:/, Resource=,namespace:default,name:pvc-b75efd69-be9e-4161-8984-4bfc432d8e6b), default "pvc-b75efd69-be9e-4161-8984-4bfc432d8e6b" is forbidden: User "system:serviceaccount:kubeedge:cloudcore" cannot get resource "default" in API group "" at the cluster scope
E0114 08:12:12.780195       1 objectsync.go:36] failed to get obj(gvr:/, Resource=,namespace:default,name:csi-hostpath-pvc), default "csi-hostpath-pvc" is forbidden: User "system:serviceaccount:kubeedge:cloudcore" cannot get resource "default" in API group "" at the cluster scope
E0114 08:12:12.784269       1 objectsync.go:36] failed to get obj(gvr:/, Resource=,namespace:default,name:10.136.77.18), default "10.136.77.18" is forbidden: User "system:serviceaccount:kubeedge:cloudcore" cannot get resource "default" in API group "" at the cluster scope
E0114 08:12:12.799033       1 objectsync.go:36] failed to get obj(gvr:/, Resource=,namespace:default,name:csi-hostpath-pvc-2), default "csi-hostpath-pvc-2" is forbidden: User "system:serviceaccount:kubeedge:cloudcore" cannot get resource "default" in API group "" at the cluster scope
E0114 08:12:12.802694       1 objectsync.go:36] failed to get obj(gvr:/, Resource=,namespace:default,name:pvc-d65b8fec-ccb3-4c38-9320-681f873a4edc), default "pvc-d65b8fec-ccb3-4c38-9320-681f873a4edc" is forbidden: User "system:serviceaccount:kubeedge:cloudcore" cannot get resource "default" in API group "" at the cluster scope
E0114 08:12:17.808242       1 objectsync.go:36] failed to get obj(gvr:/, Resource=,namespace:default,name:pvc-d65b8fec-ccb3-4c38-9320-681f873a4edc), default "pvc-d65b8fec-ccb3-4c38-9320-681f873a4edc" is forbidden: User "system:serviceaccount:kubeedge:cloudcore" cannot get resource "default" in API group "" at the cluster scope
E0114 08:12:17.811504       1 objectsync.go:36] failed to get obj(gvr:/, Resource=,namespace:default,name:10.136.77.18), default "10.136.77.18" is forbidden: User "system:serviceaccount:kubeedge:cloudcore" cannot get resource "default" in API group "" at the cluster scope
E0114 08:12:17.823988       1 objectsync.go:36] failed to get obj(gvr:/, Resource=,namespace:default,name:csi-hostpath-pvc-2), default "csi-hostpath-pvc-2" is forbidden: User "system:serviceaccount:kubeedge:cloudcore" cannot get resource "default" in API group "" at the cluster scope
E0114 08:12:17.836546       1 objectsync.go:36] failed to get obj(gvr:/, Resource=,namespace:default,name:csi-hostpath-pvc), default "csi-hostpath-pvc" is forbidden: User "system:serviceaccount:kubeedge:cloudcore" cannot get resource "default" in API group "" at the cluster scope
E0114 08:12:17.840399       1 objectsync.go:36] failed to get obj(gvr:/, Resource=,namespace:default,name:pvc-b75efd69-be9e-4161-8984-4bfc432d8e6b), default "pvc-b75efd69-be9e-4161-8984-4bfc432d8e6b" is forbidden: User "system:serviceaccount:kubeedge:cloudcore" cannot get resource "default" in API group "" at the cluster scope
E0114 08:12:22.862918       1 objectsync.go:36] failed to get obj(gvr:/, Resource=,namespace:default,name:csi-hostpath-pvc), default "csi-hostpath-pvc" is forbidden: User "system:serviceaccount:kubeedge:cloudcore" cannot get resource "default" in API group "" at the cluster scope
E0114 08:12:22.863746       1 objectsync.go:36] failed to get obj(gvr:/, Resource=,namespace:default,name:pvc-b75efd69-be9e-4161-8984-4bfc432d8e6b), default "pvc-b75efd69-be9e-4161-8984-4bfc432d8e6b" is forbidden: User "system:serviceaccount:kubeedge:cloudcore" cannot get resource "default" in API group "" at the cluster scope
E0114 08:12:22.868850       1 objectsync.go:36] failed to get obj(gvr:/, Resource=,namespace:default,name:csi-hostpath-pvc-2), default "csi-hostpath-pvc-2" is forbidden: User "system:serviceaccount:kubeedge:cloudcore" cannot get resource "default" in API group "" at the cluster scope
E0114 08:12:22.869716       1 objectsync.go:36] failed to get obj(gvr:/, Resource=,namespace:default,name:pvc-d65b8fec-ccb3-4c38-9320-681f873a4edc), default "pvc-d65b8fec-ccb3-4c38-9320-681f873a4edc" is forbidden: User "system:serviceaccount:kubeedge:cloudcore" cannot get resource "default" in API group "" at the cluster scope
E0114 08:12:22.870755       1 objectsync.go:36] failed to get obj(gvr:/, Resource=,namespace:default,name:10.136.77.18), default "10.136.77.18" is forbidden: User "system:serviceaccount:kubeedge:cloudcore" cannot get resource "default" in API group "" at the cluster scope
```

- ` k logs csi-hostpath-controller-0 csi-provisioner`:

```
I0114 08:10:48.852689       1 event.go:255] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"default", Name:"csi-hostpath-pvc", UID:"b75efd69-be9e-4161-8984-4bfc432d8e6b", APIVersion:"v1", ResourceVersion:"1967152", FieldPath:""}): type: 'Normal' reason: 'Provisioning' External provisioner is provisioning volume for claim "default/csi-hostpath-pvc"
I0114 08:10:48.852992       1 controller.go:1199] provision "default/csi-hostpath-pvc-2" class "csi-hostpath-sc": started
I0114 08:10:48.857939       1 controller.go:494] CreateVolumeRequest {Name:pvc-b75efd69-be9e-4161-8984-4bfc432d8e6b CapacityRange:required_bytes:104857600  VolumeCapabilities:[mount:<fs_type:"ext4" > access_mode:<mode:SINGLE_NODE_WRITER > ] Parameters:map[] Secrets:map[] VolumeContentSource:<nil> AccessibilityRequirements:requisite:<segments:<key:"csi.kubeedge.io/nodeid" value:"10.136.77.18" > segments:<key:"topology.hostpath.csi/node" value:"" > > preferred:<segments:<key:"csi.kubeedge.io/nodeid" value:"10.136.77.18" > segments:<key:"topology.hostpath.csi/node" value:"" > >  XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
I0114 08:10:48.858237       1 connection.go:180] GRPC call: /csi.v1.Controller/CreateVolume
I0114 08:10:48.858249       1 connection.go:181] GRPC request: {"accessibility_requirements":{"preferred":[{"segments":{"csi.kubeedge.io/nodeid":"10.136.77.18","topology.hostpath.csi/node":""}}],"requisite":[{"segments":{"csi.kubeedge.io/nodeid":"10.136.77.18","topology.hostpath.csi/node":""}}]},"capacity_range":{"required_bytes":104857600},"name":"pvc-b75efd69-be9e-4161-8984-4bfc432d8e6b","volume_capabilities":[{"AccessType":{"Mount":{"fs_type":"ext4"}},"access_mode":{"mode":1}}]}
I0114 08:10:48.867650       1 event.go:255] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"default", Name:"csi-hostpath-pvc-2", UID:"d65b8fec-ccb3-4c38-9320-681f873a4edc", APIVersion:"v1", ResourceVersion:"1967155", FieldPath:""}): type: 'Normal' reason: 'Provisioning' External provisioner is provisioning volume for claim "default/csi-hostpath-pvc-2"
I0114 08:10:48.870118       1 controller.go:494] CreateVolumeRequest {Name:pvc-d65b8fec-ccb3-4c38-9320-681f873a4edc CapacityRange:required_bytes:104857600  VolumeCapabilities:[mount:<fs_type:"ext4" > access_mode:<mode:SINGLE_NODE_WRITER > ] Parameters:map[] Secrets:map[] VolumeContentSource:<nil> AccessibilityRequirements:requisite:<segments:<key:"csi.kubeedge.io/nodeid" value:"10.136.77.18" > segments:<key:"topology.hostpath.csi/node" value:"" > > preferred:<segments:<key:"csi.kubeedge.io/nodeid" value:"10.136.77.18" > segments:<key:"topology.hostpath.csi/node" value:"" > >  XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
I0114 08:10:48.870246       1 connection.go:180] GRPC call: /csi.v1.Controller/CreateVolume
I0114 08:10:48.870252       1 connection.go:181] GRPC request: {"accessibility_requirements":{"preferred":[{"segments":{"csi.kubeedge.io/nodeid":"10.136.77.18","topology.hostpath.csi/node":""}}],"requisite":[{"segments":{"csi.kubeedge.io/nodeid":"10.136.77.18","topology.hostpath.csi/node":""}}]},"capacity_range":{"required_bytes":104857600},"name":"pvc-d65b8fec-ccb3-4c38-9320-681f873a4edc","volume_capabilities":[{"AccessType":{"Mount":{"fs_type":"ext4"}},"access_mode":{"mode":1}}]}
I0114 08:11:04.553344       1 connection.go:183] GRPC response: {"volume":{"capacity_bytes":104857600,"volume_context":{"topology.csidriver.kubeedge/node-id":"10.136.77.18"},"volume_id":"84db2f2c-7511-11ec-9503-002b67e6b46c"}}
I0114 08:11:04.553986       1 connection.go:184] GRPC error: <nil>
I0114 08:11:04.554001       1 controller.go:549] create volume rep: {CapacityBytes:104857600 VolumeId:84db2f2c-7511-11ec-9503-002b67e6b46c VolumeContext:map[topology.csidriver.kubeedge/node-id:10.136.77.18] ContentSource:<nil> AccessibleTopology:[] XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
I0114 08:11:04.554032       1 controller.go:631] successfully created PV pvc-b75efd69-be9e-4161-8984-4bfc432d8e6b for PVC csi-hostpath-pvc and csi volume name 84db2f2c-7511-11ec-9503-002b67e6b46c
I0114 08:11:04.554044       1 controller.go:647] successfully created PV {GCEPersistentDisk:nil AWSElasticBlockStore:nil HostPath:nil Glusterfs:nil NFS:nil RBD:nil ISCSI:nil Cinder:nil CephFS:nil FC:nil Flocker:nil FlexVolume:nil AzureFile:nil VsphereVolume:nil Quobyte:nil AzureDisk:nil PhotonPersistentDisk:nil PortworxVolume:nil ScaleIO:nil Local:nil StorageOS:nil CSI:&CSIPersistentVolumeSource{Driver:csi-hostpath,VolumeHandle:84db2f2c-7511-11ec-9503-002b67e6b46c,ReadOnly:false,FSType:ext4,VolumeAttributes:map[string]string{storage.kubernetes.io/csiProvisionerIdentity: 1642147832574-8081-csi-hostpath,topology.csidriver.kubeedge/node-id: 10.136.77.18,},ControllerPublishSecretRef:nil,NodeStageSecretRef:nil,NodePublishSecretRef:nil,ControllerExpandSecretRef:nil,}}
I0114 08:11:04.554197       1 controller.go:1281] provision "default/csi-hostpath-pvc" class "csi-hostpath-sc": volume "pvc-b75efd69-be9e-4161-8984-4bfc432d8e6b" provisioned
I0114 08:11:04.554220       1 controller.go:1298] provision "default/csi-hostpath-pvc" class "csi-hostpath-sc": succeeded
I0114 08:11:04.554233       1 volume_store.go:154] Saving volume pvc-b75efd69-be9e-4161-8984-4bfc432d8e6b
I0114 08:11:04.559465       1 volume_store.go:157] Volume pvc-b75efd69-be9e-4161-8984-4bfc432d8e6b saved
I0114 08:11:04.559549       1 controller.go:1014] Claim processing succeeded, removing PVC b75efd69-be9e-4161-8984-4bfc432d8e6b from claims in progress
I0114 08:11:04.561756       1 event.go:255] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"default", Name:"csi-hostpath-pvc", UID:"b75efd69-be9e-4161-8984-4bfc432d8e6b", APIVersion:"v1", ResourceVersion:"1967152", FieldPath:""}): type: 'Normal' reason: 'ProvisioningSucceeded' Successfully provisioned volume pvc-b75efd69-be9e-4161-8984-4bfc432d8e6b
I0114 08:11:29.567214       1 connection.go:183] GRPC response: {"volume":{"capacity_bytes":104857600,"volume_context":{"topology.csidriver.kubeedge/node-id":"10.136.77.18"},"volume_id":"93c3f8fd-7511-11ec-9503-002b67e6b46c"}}
I0114 08:11:29.567813       1 connection.go:184] GRPC error: <nil>
I0114 08:11:29.567828       1 controller.go:549] create volume rep: {CapacityBytes:104857600 VolumeId:93c3f8fd-7511-11ec-9503-002b67e6b46c VolumeContext:map[topology.csidriver.kubeedge/node-id:10.136.77.18] ContentSource:<nil> AccessibleTopology:[] XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
I0114 08:11:29.567857       1 controller.go:631] successfully created PV pvc-d65b8fec-ccb3-4c38-9320-681f873a4edc for PVC csi-hostpath-pvc-2 and csi volume name 93c3f8fd-7511-11ec-9503-002b67e6b46c
I0114 08:11:29.567867       1 controller.go:647] successfully created PV {GCEPersistentDisk:nil AWSElasticBlockStore:nil HostPath:nil Glusterfs:nil NFS:nil RBD:nil ISCSI:nil Cinder:nil CephFS:nil FC:nil Flocker:nil FlexVolume:nil AzureFile:nil VsphereVolume:nil Quobyte:nil AzureDisk:nil PhotonPersistentDisk:nil PortworxVolume:nil ScaleIO:nil Local:nil StorageOS:nil CSI:&CSIPersistentVolumeSource{Driver:csi-hostpath,VolumeHandle:93c3f8fd-7511-11ec-9503-002b67e6b46c,ReadOnly:false,FSType:ext4,VolumeAttributes:map[string]string{storage.kubernetes.io/csiProvisionerIdentity: 1642147832574-8081-csi-hostpath,topology.csidriver.kubeedge/node-id: 10.136.77.18,},ControllerPublishSecretRef:nil,NodeStageSecretRef:nil,NodePublishSecretRef:nil,ControllerExpandSecretRef:nil,}}
I0114 08:11:29.567918       1 controller.go:1281] provision "default/csi-hostpath-pvc-2" class "csi-hostpath-sc": volume "pvc-d65b8fec-ccb3-4c38-9320-681f873a4edc" provisioned
I0114 08:11:29.567937       1 controller.go:1298] provision "default/csi-hostpath-pvc-2" class "csi-hostpath-sc": succeeded
I0114 08:11:29.567941       1 volume_store.go:154] Saving volume pvc-d65b8fec-ccb3-4c38-9320-681f873a4edc
I0114 08:11:29.570947       1 volume_store.go:157] Volume pvc-d65b8fec-ccb3-4c38-9320-681f873a4edc saved
I0114 08:11:29.570995       1 controller.go:1014] Claim processing succeeded, removing PVC d65b8fec-ccb3-4c38-9320-681f873a4edc from claims in progress
I0114 08:11:29.571025       1 event.go:255] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"default", Name:"csi-hostpath-pvc-2", UID:"d65b8fec-ccb3-4c38-9320-681f873a4edc", APIVersion:"v1", ResourceVersion:"1967155", FieldPath:""}): type: 'Normal' reason: 'ProvisioningSucceeded' Successfully provisioned volume pvc-d65b8fec-ccb3-4c38-9320-681f873a4edc
I0114 08:17:43.588359       1 reflector.go:389] sigs.k8s.io/sig-storage-lib-external-provisioner/controller/controller.go:807: Watch close - *v1.StorageClass total 2 items received
I0114 08:18:51.587927       1 reflector.go:389] sigs.k8s.io/sig-storage-lib-external-provisioner/controller/controller.go:801: Watch close - *v1.PersistentVolumeClaim total 11 items received
```

- ` k logs csi-hostpath-controller-0 csi-driver`: 

```
I0114 08:10:32.663119       1 utils.go:137] grpc response: {"capabilities":[{"Type":{"Rpc":{"type":1}}},{"Type":{"Rpc":{"type":2}}}]}
I0114 08:10:32.664606       1 utils.go:131] grpc call: /csi.v1.Controller/ControllerGetCapabilities
I0114 08:10:32.664616       1 utils.go:132] grpc request: {}
I0114 08:10:32.664639       1 utils.go:137] grpc response: {"capabilities":[{"Type":{"Rpc":{"type":1}}},{"Type":{"Rpc":{"type":2}}}]}
I0114 08:10:48.866186       1 utils.go:131] grpc call: /csi.v1.Controller/CreateVolume
I0114 08:10:48.866223       1 utils.go:132] grpc request: {"accessibility_requirements":{"preferred":[{"segments":{"csi.kubeedge.io/nodeid":"10.136.77.18","topology.hostpath.csi/node":""}}],"requisite":[{"segments":{"csi.kubeedge.io/nodeid":"10.136.77.18","topology.hostpath.csi/node":""}}]},"capacity_range":{"required_bytes":104857600},"name":"pvc-b75efd69-be9e-4161-8984-4bfc432d8e6b","volume_capabilities":[{"AccessType":{"Mount":{"fs_type":"ext4"}},"access_mode":{"mode":1}}]}
I0114 08:10:48.866656       1 controllerserver.go:73] create volume request id=pvc-b75efd69-be9e-4161-8984-4bfc432d8e6b req=&csi.CreateVolumeRequest{Name:"pvc-b75efd69-be9e-4161-8984-4bfc432d8e6b", CapacityRange:(*csi.CapacityRange)(0xc000097a70), VolumeCapabilities:[]*csi.VolumeCapability{(*csi.VolumeCapability)(0xc0004a11c0)}, Parameters:map[string]string(nil), Secrets:map[string]string(nil), VolumeContentSource:(*csi.VolumeContentSource)(nil), AccessibilityRequirements:(*csi.TopologyRequirement)(0xc0004c1a40), XXX_NoUnkeyedLiteral:struct {}{}, XXX_unrecognized:[]uint8(nil), XXX_sizecache:0}
I0114 08:10:48.866737       1 controllerserver.go:83] !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! :requisite:<segments:<key:"csi.kubeedge.io/nodeid" value:"10.136.77.18" > segments:<key:"topology.hostpath.csi/node" value:"" > > preferred:<segments:<key:"csi.kubeedge.io/nodeid" value:"10.136.77.18" > segments:<key:"topology.hostpath.csi/node" value:"" > > 
I0114 08:10:48.866791       1 controllerserver.go:122] topology requirements: %#vrequisite:<segments:<key:"csi.kubeedge.io/nodeid" value:"10.136.77.18" > segments:<key:"topology.hostpath.csi/node" value:"" > > preferred:<segments:<key:"csi.kubeedge.io/nodeid" value:"10.136.77.18" > segments:<key:"topology.hostpath.csi/node" value:"" > > 
I0114 08:10:48.866818       1 controllerserver.go:89] picked edge node: 10.136.77.18
I0114 08:10:48.872740       1 utils.go:131] grpc call: /csi.v1.Controller/CreateVolume
I0114 08:10:48.872756       1 utils.go:132] grpc request: {"accessibility_requirements":{"preferred":[{"segments":{"csi.kubeedge.io/nodeid":"10.136.77.18","topology.hostpath.csi/node":""}}],"requisite":[{"segments":{"csi.kubeedge.io/nodeid":"10.136.77.18","topology.hostpath.csi/node":""}}]},"capacity_range":{"required_bytes":104857600},"name":"pvc-d65b8fec-ccb3-4c38-9320-681f873a4edc","volume_capabilities":[{"AccessType":{"Mount":{"fs_type":"ext4"}},"access_mode":{"mode":1}}]}
I0114 08:10:48.872939       1 controllerserver.go:73] create volume request id=pvc-d65b8fec-ccb3-4c38-9320-681f873a4edc req=&csi.CreateVolumeRequest{Name:"pvc-d65b8fec-ccb3-4c38-9320-681f873a4edc", CapacityRange:(*csi.CapacityRange)(0xc0000fc030), VolumeCapabilities:[]*csi.VolumeCapability{(*csi.VolumeCapability)(0xc0004a1600)}, Parameters:map[string]string(nil), Secrets:map[string]string(nil), VolumeContentSource:(*csi.VolumeContentSource)(nil), AccessibilityRequirements:(*csi.TopologyRequirement)(0xc0000763c0), XXX_NoUnkeyedLiteral:struct {}{}, XXX_unrecognized:[]uint8(nil), XXX_sizecache:0}
I0114 08:10:48.872962       1 controllerserver.go:83] !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! :requisite:<segments:<key:"csi.kubeedge.io/nodeid" value:"10.136.77.18" > segments:<key:"topology.hostpath.csi/node" value:"" > > preferred:<segments:<key:"csi.kubeedge.io/nodeid" value:"10.136.77.18" > segments:<key:"topology.hostpath.csi/node" value:"" > > 
I0114 08:10:48.872996       1 controllerserver.go:122] topology requirements: %#vrequisite:<segments:<key:"csi.kubeedge.io/nodeid" value:"10.136.77.18" > segments:<key:"topology.hostpath.csi/node" value:"" > > preferred:<segments:<key:"csi.kubeedge.io/nodeid" value:"10.136.77.18" > segments:<key:"topology.hostpath.csi/node" value:"" > > 
I0114 08:10:48.873015       1 controllerserver.go:89] picked edge node: 10.136.77.18
I0114 08:11:04.552889       1 controllerserver.go:104] create volume response: &csi.CreateVolumeResponse{Volume:(*csi.Volume)(0xc000393080), XXX_NoUnkeyedLiteral:struct {}{}, XXX_unrecognized:[]uint8(nil), XXX_sizecache:0}
I0114 08:11:04.552982       1 controllerserver.go:117] returning volume response: &csi.CreateVolumeResponse{Volume:(*csi.Volume)(0xc000393380), XXX_NoUnkeyedLiteral:struct {}{}, XXX_unrecognized:[]uint8(nil), XXX_sizecache:0}
I0114 08:11:04.552994       1 utils.go:137] grpc response: {"volume":{"capacity_bytes":104857600,"volume_context":{"topology.csidriver.kubeedge/node-id":"10.136.77.18"},"volume_id":"84db2f2c-7511-11ec-9503-002b67e6b46c"}}
I0114 08:11:29.566889       1 controllerserver.go:104] create volume response: &csi.CreateVolumeResponse{Volume:(*csi.Volume)(0xc0003936e0), XXX_NoUnkeyedLiteral:struct {}{}, XXX_unrecognized:[]uint8(nil), XXX_sizecache:0}
I0114 08:11:29.566923       1 controllerserver.go:117] returning volume response: &csi.CreateVolumeResponse{Volume:(*csi.Volume)(0xc000393860), XXX_NoUnkeyedLiteral:struct {}{}, XXX_unrecognized:[]uint8(nil), XXX_sizecache:0}
I0114 08:11:29.566934       1 utils.go:137] grpc response: {"volume":{"capacity_bytes":104857600,"volume_context":{"topology.csidriver.kubeedge/node-id":"10.136.77.18"},"volume_id":"93c3f8fd-7511-11ec-9503-002b67e6b46c"}}
```


- pvc && pod 

```
[root@cloud cloud]# k get pvc
NAME                 STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS      AGE
csi-hostpath-pvc     Bound    pvc-b75efd69-be9e-4161-8984-4bfc432d8e6b   100Mi      RWO            csi-hostpath-sc   10m
csi-hostpath-pvc-2   Bound    pvc-d65b8fec-ccb3-4c38-9320-681f873a4edc   100Mi      RWO            csi-hostpath-sc   10m
[root@cloud cloud]# k get pods
NAME                        READY   STATUS    RESTARTS   AGE
csi-hostpath-controller-0   3/3     Running   0          11m
csi-hostpath-edge-rhw6h     2/2     Running   0          11m
testpod-1                   0/1     Pending   0          10m
testpod-2                   0/1     Pending   0          10m
[root@cloud cloud]# k get sc
NAME              PROVISIONER    RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
csi-hostpath-sc   csi-hostpath   Retain          WaitForFirstConsumer   false                  11m
```

**Comment by moolen** [comment]
hey, this is an old branch, integration tested from this branch: https://github.com/moolen/kubeedge/tree/wip/topo-integration it contains all the changes needed to get it running.

I created this gist with a stripped down version of the manifests i use for testing: https://gist.github.com/moolen/4313754c376d3f81d5e15147463aea85

Also don't forget to label the nodes correctly 
 
```yaml
apiVersion: v1
kind: Node
metadata:
  labels:
    csi.kubeedge.io/nodeid: edge-1 # required for this feature MUST BE ADDED MANUALLY for now
    topology.hostpath.csi/node: edge-1 # required for this particular driver
  name: edge-1
```
Also note that a CSINode object should be created by edgecore. This may take a minute or two.
```yaml
apiVersion: storage.k8s.io/v1
kind: CSINode
metadata:
  name: edge-1
spec:
  drivers:
  - name: csi-hostpath
    nodeID: edge-1
    topologyKeys:
    - csi.kubeedge.io/nodeid
    - topology.hostpath.csi/node
```

**Comment by shumintao** [comment]
I seems to have a permission problem


- csi-hostpath-controller
![image](https://user-images.githubusercontent.com/26895243/149704992-37d589f5-7991-4f90-b0d0-6e921036dad2.png)

- cloudcore

![image](https://user-images.githubusercontent.com/26895243/149705009-9e2a0fdf-3228-40e9-8736-06e6c2d749b2.png)

**Comment by moolen** [comment]
I think the error you mentioned is unrelated (i see those, too). However the `csi-hostpath-controller` has a RBAC issue. The setup is in [their repo](https://github.com/kubernetes-csi/csi-driver-host-path/blob/master/deploy/kubernetes-1.21/hostpath/csi-hostpath-plugin.yaml). Feel Free to hit me up in Kubedge or Kubernetes slack if you have any issues.

**Comment by shumintao** [comment]
Thank you for your reply, I found that there's an error in [https://github.com/moolen/kubeedge/tree/wip/topo-integration](https://github.com/moolen/kubeedge/tree/wip/topo-integration) branch,  I repaired the two places can run at present：

- edged.go
```
		case constants.CSIResourceTypeVolume:
			klog.Infof("volume operation type=%s content=%#v", op, result.Content)
			cb, ok := result.GetContent().(string)
			if !ok {
				klog.Errorf("unexpected volume content: %#v, expected string", content)
				beehiveContext.SendResp(*model.NewErrorMessage(&result, err.Error()))
				continue
			}
			res, err := e.handleVolume(op, []byte(cb))
			if err != nil {
				klog.Errorf("handle volume failed: %v", err)
				beehiveContext.SendResp(*model.NewErrorMessage(&result, err.Error()))
				continue
			}
			js, err := (&jsonpb.Marshaler{}).MarshalToString(proto.MessageV1(res))
			if err != nil {
				klog.Error("failed to marshal handle volume response")
				beehiveContext.SendResp(*model.NewErrorMessage(&result, err.Error()))
				continue
			}
			resp := result.NewRespByMessage(&result, js)
			klog.Infof("handle volume sending response to cloud content=%s resp=%#v", js, resp)
			beehiveContext.SendResp(*resp)
```

- 03-clusterrole.yaml

```
- apiGroups: [""]
  resources: ["pods"]
  verbs: ["create","delete","get","list","patch","update","watch"]
```

**Comment by moolen** [comment]
Thanks @shumintao, i completely forgot about that :sweat_smile: that was part of #3458.

**Comment by moolen** [comment]
As discussed in the community meeting today, we want to explore the alternative of running all csi related drivers (provisioner, attacher, driver...) on the edge node. These components require access to the control plane - that can be implemented using edgemesh.

**Comment by fisherxu** [comment]
> As discussed in the community meeting today, we want to explore the alternative of running all csi related drivers (provisioner, attacher, driver...) on the edge node. These components require access to the control plane - that can be implemented using edgemesh.

We can use the metaServer in edgecore to connect to the control plane, edgemesh also use this server to talk to the control plane.
This issue explains how to use the metaServer to enable the flannel which needs to talk to the control plane running in th edge.
https://github.com/kubeedge/kubeedge/issues/2677

I'll also write some docs for it.  @moolen

**Comment by moolen** [comment]
I stitched together a [example gist](https://gist.github.com/moolen/796a257a079e472b6925996eda148356) that someone can use as a reference point (e.g. for documentation). 

However i have a problem with current master branch:

```
Observed a panic: "invalid memory address or nil poi>
goroutine 1663 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x23847a0, 0x3fc23f0})
        /home/ubuntu/kubeedge/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0x85
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x30})
        /home/ubuntu/kubeedge/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x75
panic({0x23847a0, 0x3fc23f0})
        /usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/kubeedge/kubeedge/edge/pkg/edged/volume/csi.(*csiMountMgr).GetMetrics(0xc0012b0f00)
        <autogenerated>:1 +0x29
[...]
```

This is fixed with https://github.com/moolen/kubeedge/commit/c9194f88e5358dc3a88a85307fe5ddf93c9c667f.

I'll talk to my client and check if distributed provisioning is an option for them.

**Comment by moolen** [comment]
Distributed provisioning works for my client, once #3581 is merged we can close this issue. We'll do some thorough testing and see if any other issues arise from that approach.
